File: Packages/jp.shiranui-isuzu.unity-mcp/Samples~/UnityMCPHandlerSamplesJS/AssetSearchCommandHandler.js
import { z } from "zod";
import { BaseCommandHandler } from "../core/BaseCommandHandler.js";
/**
 * Command handler for searching assets in the Unity project.
 */
export class AssetSearchCommandHandler extends BaseCommandHandler {
    /**
     * Gets the command prefix for this handler.
     */
    get commandPrefix() {
        return "asset";
    }
    /**
     * Gets the description of this command handler.
     */
    get description() {
        return "Search and retrieve assets from the project";
    }
    /**
     * Executes the command with the given parameters.
     * @param action The action to execute.
     * @param parameters The parameters for the command.
     * @returns A Promise that resolves to a JSON object containing the execution result.
     */
    async execute(action, parameters) {
        switch (action.toLowerCase()) {
            case "search":
                return this.searchAssets(parameters);
            case "findbyname":
                return this.findByName(parameters);
            case "findbytype":
                return this.findByType(parameters);
            default:
                return {
                    success: false,
                    error: `Unknown action: ${action}. Supported actions: search, findByName, findByType`
                };
        }
    }
    /**
     * Gets the tool definitions supported by this handler.
     * @returns A map of tool names to their definitions.
     */
    getToolDefinitions() {
        const tools = new Map();
        // Add asset_search tool
        tools.set("asset_search", {
            description: "Search for assets in the Unity project using a query",
            parameterSchema: {
                query: z.string().describe("The search query"),
                limit: z.number().optional().describe("Maximum number of results to return (default: 100)")
            },
            annotations: {
                title: "Search Assets",
                readOnlyHint: true,
                destructiveHint: false,
                idempotentHint: true,
                openWorldHint: false
            }
        });
        // Add asset_findByName tool
        tools.set("asset_findByName", {
            description: "Find assets by name in the Unity project",
            parameterSchema: {
                name: z.string().describe("The name to search for"),
                exact: z.boolean().optional().describe("Whether to perform an exact match (default: false)"),
                limit: z.number().optional().describe("Maximum number of results to return (default: 100)")
            },
            annotations: {
                title: "Find Assets by Name",
                readOnlyHint: true,
                destructiveHint: false,
                idempotentHint: true,
                openWorldHint: false
            }
        });
        // Add asset_findByType tool
        tools.set("asset_findByType", {
            description: "Find assets by type in the Unity project",
            parameterSchema: {
                type: z.string().describe("The type name to search for (e.g., 'Texture2D', 'Material')"),
                limit: z.number().optional().describe("Maximum number of results to return (default: 100)")
            },
            annotations: {
                title: "Find Assets by Type",
                readOnlyHint: true,
                destructiveHint: false,
                idempotentHint: true,
                openWorldHint: false
            }
        });
        return tools;
    }
    /**
     * Searches for assets using a query.
     * @param parameters Parameters containing the search query.
     * @returns A Promise that resolves to a JSON object containing the search results.
     */
    async searchAssets(parameters) {
        try {
            // First ensure we have a valid connection to Unity
            await this.ensureUnityConnection();
            // Forward the request to Unity
            return await this.sendUnityRequest(`${this.commandPrefix}.search`, parameters);
        }
        catch (ex) {
            const errorMessage = ex instanceof Error ? ex.message : String(ex);
            console.error(`Error searching assets: ${errorMessage}`);
            return {
                success: false,
                error: errorMessage
            };
        }
    }
    /**
     * Finds assets by name.
     * @param parameters Parameters containing the name to search for.
     * @returns A Promise that resolves to a JSON object containing the search results.
     */
    async findByName(parameters) {
        try {
            // First ensure we have a valid connection to Unity
            await this.ensureUnityConnection();
            // Forward the request to Unity
            return await this.sendUnityRequest(`${this.commandPrefix}.findByName`, parameters);
        }
        catch (ex) {
            const errorMessage = ex instanceof Error ? ex.message : String(ex);
            console.error(`Error finding assets by name: ${errorMessage}`);
            return {
                success: false,
                error: errorMessage
            };
        }
    }
    /**
     * Finds assets by type.
     * @param parameters Parameters containing the type to search for.
     * @returns A Promise that resolves to a JSON object containing the search results.
     */
    async findByType(parameters) {
        try {
            // First ensure we have a valid connection to Unity
            await this.ensureUnityConnection();
            // Forward the request to Unity
            return await this.sendUnityRequest(`${this.commandPrefix}.findByType`, parameters);
        }
        catch (ex) {
            const errorMessage = ex instanceof Error ? ex.message : String(ex);
            console.error(`Error finding assets by type: ${errorMessage}`);
            return {
                success: false,
                error: errorMessage
            };
        }
    }
}

File: Packages/jp.shiranui-isuzu.unity-mcp/Samples~/UnityMCPHandlerSamples/Editor/AssetSearchCommandHandler.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityMCP.Editor.Core;

namespace UnityMCPHandlerSamples.Editor
{

    /// <summary>
    /// Command handler for searching assets in the Unity project.
    /// </summary>
    internal sealed class AssetSearchCommandHandler : IMcpCommandHandler
    {
        /// <summary>
        /// Gets the command prefix for this handler.
        /// </summary>
        public string CommandPrefix => "asset";

        /// <summary>
        /// Gets the description of this command handler.
        /// </summary>
        public string Description => "Search and retrieve assets from the project";

        /// <summary>
        /// Executes the command with the given parameters.
        /// </summary>
        /// <param name="action">The action to execute.</param>
        /// <param name="parameters">The parameters for the command.</param>
        /// <returns>A JSON object containing the execution result.</returns>
        public JObject Execute(string action, JObject parameters)
        {
            return action.ToLower() switch
            {
                "search" => this.SearchAssets(parameters),
                "findbyname" => this.FindAssetsByName(parameters),
                "findbytype" => this.FindAssetsByType(parameters),
                _ => new JObject { ["success"] = false, ["error"] = $"Unknown action: {action}. Supported actions are 'search', 'findbyname', and 'findbytype'." }
            };
        }

        /// <summary>
        /// Searches for assets using the provided query.
        /// </summary>
        /// <param name="parameters">The search parameters.</param>
        /// <returns>A JSON object containing the search results.</returns>
        private JObject SearchAssets(JObject parameters)
        {
            var query = parameters["query"]?.ToString();
            if (string.IsNullOrWhiteSpace(query))
            {
                return new JObject
                {
                    ["success"] = false,
                    ["error"] = "Query parameter is required"
                };
            }

            var limit = parameters["limit"]?.Value<int>() ?? 100;
            var results = AssetDatabase.FindAssets(query);

            return this.FormatSearchResults(results, limit);
        }

        /// <summary>
        /// Finds assets by name.
        /// </summary>
        /// <param name="parameters">The search parameters.</param>
        /// <returns>A JSON object containing the search results.</returns>
        private JObject FindAssetsByName(JObject parameters)
        {
            var name = parameters["name"]?.ToString();
            if (string.IsNullOrWhiteSpace(name))
            {
                return new JObject
                {
                    ["success"] = false,
                    ["error"] = "Name parameter is required"
                };
            }

            var limit = parameters["limit"]?.Value<int>() ?? 100;
            var exact = parameters["exact"]?.Value<bool>() ?? false;

            // Build the search query for name search
            var query = exact ? name : $"{name}";
            var results = AssetDatabase.FindAssets(query);

            // If exact match is requested, we need to filter the results
            if (exact)
            {
                var exactResults = new List<string>();
                foreach (var guid in results)
                {
                    var assetPath = AssetDatabase.GUIDToAssetPath(guid);
                    var assetName = System.IO.Path.GetFileNameWithoutExtension(assetPath);
                    if (string.Equals(assetName, name, StringComparison.OrdinalIgnoreCase))
                    {
                        exactResults.Add(guid);
                    }
                }

                results = exactResults.ToArray();
            }

            return this.FormatSearchResults(results, limit);
        }

        /// <summary>
        /// Finds assets by type.
        /// </summary>
        /// <param name="parameters">The search parameters.</param>
        /// <returns>A JSON object containing the search results.</returns>
        private JObject FindAssetsByType(JObject parameters)
        {
            var typeName = parameters["type"]?.ToString();
            if (string.IsNullOrWhiteSpace(typeName))
            {
                return new JObject
                {
                    ["success"] = false,
                    ["error"] = "Type parameter is required"
                };
            }

            var limit = parameters["limit"]?.Value<int>() ?? 100;

            // Build the search query for type search
            var query = $"t:{typeName}";
            var results = AssetDatabase.FindAssets(query);

            return this.FormatSearchResults(results, limit);
        }

        /// <summary>
        /// Formats the search results as a JSON object.
        /// </summary>
        /// <param name="guids">The GUIDs of the found assets.</param>
        /// <param name="limit">The maximum number of results to return.</param>
        /// <returns>A JSON object containing the formatted search results.</returns>
        private JObject FormatSearchResults(string[] guids, int limit)
        {
            var results = new JArray();
            var count = Math.Min(guids.Length, limit);

            for (int i = 0; i < count; i++)
            {
                var guid = guids[i];
                var assetPath = AssetDatabase.GUIDToAssetPath(guid);

                if (string.IsNullOrEmpty(assetPath))
                {
                    continue;
                }

                var asset = AssetDatabase.LoadAssetAtPath<UnityEngine.Object>(assetPath);
                if (asset == null)
                {
                    continue;
                }

                var assetName = System.IO.Path.GetFileNameWithoutExtension(assetPath);
                var assetType = asset.GetType().Name;

                results.Add(new JObject
                {
                    ["guid"] = guid,
                    ["path"] = assetPath,
                    ["name"] = assetName,
                    ["type"] = assetType
                });
            }

            return new JObject
            {
                ["success"] = true,
                ["count"] = results.Count,
                ["total"] = guids.Length,
                ["results"] = results
            };
        }
    }
}
